assignment8/sql_intro.py

```diff
@@ -0,0 +1,161 @@
+def add_publisher(cursor, publisher_name):
+    try:
+        cursor.execute(
+            "INSERT INTO publishers (publisher_name) VALUES (?)", (publisher_name,))
+    except sqlite3.IntegrityError:
+        print(f"{publisher_name} is already in the database.")
+
+
+def add_magazine(cursor, magazine_name, publisher_id):
+    try:
+        cursor.execute(
+            "INSERT INTO magazines (magazine_name, publisher_id) VALUES (?, ?)", (magazine_name, publisher_id))
+    except sqlite3.IntegrityError:
+        print(f"{magazine_name} is already in the database.")
+
+
+def add_subscriber(cursor, subscriber_name, address):
+    cursor.execute(
+        "SELECT * FROM subscribers WHERE subscriber_name = ? AND address = ?", (subscriber_name, address))
+    if cursor.fetchone():
+        print(f"{subscriber_name} at {address} already exist.")
+        return
+    cursor.execute(
+        "INSERT INTO subscribers (subscriber_name, address) VALUES (?,?)", 
+        (subscriber_name, address))
+
+
+def register_subscription(cursor, subscriber, magazine, expiration_date):
+    cursor.execute(
+        "SELECT subscriber_id FROM subscribers WHERE subscriber_name = ?", 
+        (subscriber,))
+    results = cursor.fetchone()
+    if results:
+        subscriber_id = results[0]
+    else:
+        print(f"There was no subscriber named {subscriber}.")
+        return
+    
+
+    cursor.execute(
+        "SELECT magazine_id FROM magazines WHERE magazine_name = ?", 
+        (magazine,))
+    results = cursor.fetchone()
+    if results:
+        magazine_id = results[0]
+    else:
+        print(f"There was no magazine named {magazine}.")
+        return
+    
+
+    # Check if this subscription already exists
+    cursor.execute(
+        "SELECT * FROM subscriptions WHERE subscriber_id = ? AND magazine_id = ?",
+        (subscriber_id, magazine_id)
+    )
+    if cursor.fetchone():
+        print(f"Subscriber {subscriber} already subscribed to {magazine}.")
+        return
+
+    # If not, insert the new subscription
+    cursor.execute(
+        "INSERT INTO subscriptions (subscriber_id, magazine_id, expiration_date) VALUES (?, ?, ?)", 
+        (subscriber_id, magazine_id, expiration_date))
+
+
+# Task 1: Create a New SQLite Database
+try:
+    with sqlite3.connect('../db/magazines.db') as conn:
+        conn.execute("PRAGMA foreign_keys = 1")
+        cursor = conn.cursor()
+
+        # Task 2: Define Database Structure
+        cursor.execute("""
+        CREATE TABLE IF NOT EXISTS publishers (
+            publisher_id INTEGER PRIMARY KEY,
+            publisher_name TEXT NOT NULL UNIQUE
+        )
+        """)
+
+        cursor.execute("""
+        CREATE TABLE IF NOT EXISTS magazines (
+            magazine_id INTEGER PRIMARY KEY,
+            magazine_name TEXT NOT NULL UNIQUE,
+            publisher_id INTEGER NOT NULL,
+            FOREIGN KEY (publisher_id) REFERENCES publishers(publisher_id) ON UPDATE CASCADE
```

Comment:
Good job defining the database structure and setting up primary/foreign keys. Using ON UPDATE CASCADE is a really nice touch to help maintain integrity automatically if the parent key ever changes.